common/include/new

```diff
@@ -0,0 +1,29 @@
+#ifndef __NEW_H__
+#define __NEW_H__
+
+#include <cstddef>
+
+namespace std {
+struct nothrow_t {
+  explicit nothrow_t() = default;
+};
+
+extern const nothrow_t nothrow;
+
+typedef void (*new_handler)();
+
+new_handler get_new_handler() noexcept;
```

Comment:
I'm conflicted about the value of SJLJ exception handlng.  That design has a runtime cost and gave a lot of programmers a bad impression about exception support. SJLJ is arguably worse than simply not having exception support.  The modern exception implementation, using out-of-line exception tables, is not considered to be expensive at all. I have no idea why it conflicts with ELF support.  Doesn't the Linux-ELF target support exceptions?

I do assume that most users of the MOS target won't have much use for exceptions anyway; so in general I would call it a low-priority feature, but one that is ultimately required if you're going to try to get anywhere near standard compliance. It's really baked into the language at this point, and you have to write some pretty non-standard code to really avoid all the situations where a clang (or any modern implementation) will generate a throw.  For MOS; it's probably most important to just have predictable behavior in the exception-throwing situations: basically a good std::terminate() to jump to since it's unlikely that you'll ever have a good place to catch an exception. 

FWIW, compiling with -fno-exceptions absolutely does not mean your program won't generate an exception-- you'll still see exceptions thrown from within libc++ and __cxa_* functions; -fno-exceptions does nothing to change that. 

I don't know how much it matters on `std::get_new_handler`, but in general having the correct declaration of `noexcept` is increasingly important.  It's really an orthogonal issue from whether or not the code generator actually emits exception frames.  

Whether or not a function is `noexcept` is part of it's function signature, and as such it's something that plays a part in type selection for templates. See the `std::is_nothrow_*`  type_traits for examples. 

What's more I've already run into issues with the code clang++ generates for `new (std::nothrow) T{}` for type T that has a non-trivial constructor: clang++ will only check the return type from the underlying `operator new` implementation if  the `operator new(..., nothrow)` overload is marked `noexcept`.  Otherwise, it will blindly call the object constructor with `this` set to nullptr if the allocation fails!  Based on that experience, I am assuming that clang++ really does require functions implicitly defined in the standard library to correctly define non-throwing functions; otherwise bad things will occur.

Reply:
>The modern exception implementation, using out-of-line exception tables, is not considered to be expensive at all.
Very much depends on who you talk to. It's certainly 

> I have no idea why it conflicts with ELF support. Doesn't the Linux-ELF target support exceptions?
Ah, I misspoke; not ELF, but the CFI tables. It's very unlikely that the existing exception table generation code in LLVM is a practical implementation on the 6502, so it'd be a small researchey project to make a reasonable new one.

> I do assume that most users of the MOS target won't have much use for exceptions anyway; so in general I would call it a low-priority feature, but one that is ultimately required if you're going to try to get anywhere near standard compliance.

"C++11 standards compliance, except enforcing -fno-exceptions," is still a reasonable guarantee to offer to users; akin to the "C99 standards compliance, except without floats" we currently offer. We wouldn't be alone in that either: the NVPTX and AMDCPN GPU backends both operate exactly this way, to the point of simply overriding the -fexception flag immediately after it's parsed. There's an if statement in Clang there with our name on it if we want to go that route.

I do eventually want floats, just for completeness sake, but adding floats to the runtime is a considerably less invasive change than exceptions.

> FWIW, compiling with -fno-exceptions absolutely does not mean your program won't generate an exception-- you'll still see exceptions thrown from within libc++ and __cxa_* functions; -fno-exceptions does nothing to change that.

Sure it does; we're the ones writing libc++ and all the __cxa_ functions.
 
> What's more I've already run into issues with the code clang++ generates for `new (std::nothrow) T{}` for type T that has a non-trivial constructor: clang++ will only check the return type from the underlying `operator new` implementation if the `operator new(..., nothrow)` overload is marked `noexcept`. Otherwise, it will blindly call the object constructor with `this` set to nullptr if the allocation fails! Based on that experience, I am assuming that clang++ really does require functions implicitly defined in the standard library to correctly define non-throwing functions; otherwise bad things will occur.

In this case, I'd expect a `-fno-exceptions`-ey implementation of the "throwing" `new` to call `std::terminate`; in that case, the offending `this` deallocation would never be reached. The same could be done for every mandated exception in the standard library.

Reply:
> Very much depends on who you talk to. It's definitely not as bad as its worst detractors think, but it's also very uncommon in, say, the game dev world. To be fair, I learned this from a local talk by a Sony LLVM dev trying to lower the overhead and convince his company that exceptions are sorta conditionally OK now.

I would say-- ask the upstream clang developers. Game dev is a special case in that exceptions have limited usefulness even when they are free. My perspective is for programs that have to actual deal with unforseen exception circumstances exceptions are "free" in the sense that non-throwing code paths generate identical code regardless of whether exceptions are on.

> There's an if statement in Clang there with our name on it if we want to go that route.

I would say that the MOS target is more general purpose than the GPU. As such you are more likely to have users that want exceptions for their more complex, non-game, programs.  Doesn't have to be a high priority feature though.

> In this case, I'd expect a -fno-exceptions-ey implementation of the "throwing" new to call std::terminate; in that case, the offending this deallocation would never be reached. The same could be done for every mandated exception in the standard library.

I think you're missing the point on this one:  `operator new(..., nothrow)` is required to _not_ throw. If allocation fails, clang must skip the constructor call and return null. Clang++ only generates this code if your implementation of the operator function is marked noexcept.

Reply:
So, I had a quick discussion with the other project members about this, and we've reached a consensus.

While C++ exceptions do have their uses, they produce an unavoidable size overhead due to CFI tables whether or not exceptions are thrown, and are considerably slower than regular control flow when exceptions are thrown. Worse, exceptions are a leaky abstraction from a compiler author's POV; the possibility of an exception travelling through a code region on unwind requires special casing throughout the compiler. Accordingly, there's a considerable maintenance burden involved with supporting exceptions: in the runtime libraries, in the standard library, and in the compiler itself.

However, none of the above is a reason to permanently ban exception handling; just to put it off. If and when we get around to implementing proper exception handling, we should do so all in one sweep, from the compiler, through the  runtime libraries, to the standard library. This will help us better understand the maintenance costs involved and allow us to avoid paying any of these costs until then.

Accordingly, as soon as I get the chance, I'll be adding the necessary lines to force `-fno-exceptions` for the MOS target in clang.

The implications for this PR:
1. The `noexcept` specificiers for internal functions won't do anything. Please remove them; we can add them later.
2. Please also implement the "throw" versions of new as well; we can implement them up to the `-fno-excpetions` guarantee by having them call `std::terminate` in case of allocation failure.

Reply:
I'm o.k. with running no-exceptions.  I'd rather we just leave them un-implemented so that code that tries to call throw just doesn't compile, which is the current status quo.

However:
> The noexcept specificiers for internal functions won't do anything. Please remove them; we can add them later.

This is demonstrably false.  The clang++ front-end generates different, incorrect code, if the noexcept is missing on the functions that require it.  Try it yourself if you don't believe me-- remove noexcept from the operator new() overloads and build/run the hello-new example.  Whether or not you compile with fno-exceptions has zero effect on this.

>Please also implement the "throw" versions of new as well; we can implement them up to the -fno-excpetions guarantee by having them call std::terminate in case of allocation failure.

These are already implemented at, for example new.cc line 300. At this point it calls abort() rather than std::terminate.  I can implement std::terminate in a subsequent PR and will swap it over at that point.

Reply:
By `internal functions`, I mean functions internal to the new.cc file; functions like `get_block()`.

Reply:
> These are already implemented at, for example new.cc line 300. At this point it calls abort() rather than std::terminate. I can implement std::terminate in a subsequent PR and will swap it over at that point.

Sorry, I realized this immediately after posting. No worries; we can get to std::terminate later.